COS-3013: overlay node image before bootstrapping if necessary

jlebon opened the pull request:
As per https://github.com/openshift/enhancements/pull/1637, we're trying to get rid of all OpenShift-versioned components from the bootimages.

This means that there will no longer be `oc`, `kubelet`, or `crio` binaries for example, which bootstrapping obviously relies on.

Instead, now we change things up so that early on when booting the bootstrap node, we pull down the node image, unencapsulate it (this just means convert it back to an OSTree commit), then mount over its `/usr`, and import new `/etc` content.

This is done by isolating to a different systemd target to only bring up the minimum number of services to do the pivot and then carry on with bootstrapping.

This does not incur additional reboots and should be compatible with AI/ABI/SNO. But it is of course, a huge conceptual shift in how bootstrapping works. With this, we would now always be sure that we're using the same binaries as the target version as part of bootstrapping, which should alleviate some issues such as AI late-binding (see e.g. https://issues.redhat.com/browse/MGMT-16705).

The big exception of course being the kernel. Relatedly, note we do persist `/usr/lib/modules` from the booted system so that loading kernel modules still works.

To be conservative, the new logic only kicks in when using bootimages which do not have `oc`. This will allow us to ratchet this in more easily.

Down the line, we should be able to replace some of this with `bootc apply-live` once that's available (and also works in a live environment). (See https://github.com/containers/bootc/issues/76.)

For full context, see the linked enhancement and discussions there.

jlebon commented:
/test e2e-aws-ovn

jlebon commented:
/retest

jlebon commented:
> :heavy_check_mark: **ci/prow/e2e-aws-ovn** — Job succeeded
> :heavy_check_mark: **ci/prow/e2e-openstack-ovn** — Job succeeded

Sweet! :tada: 

Looks like e2e-aws-ovn-heterogeneous failed on a different issue.

jlebon commented:
/retest

jlebon commented:
/test e2e-agent-compact-ipv4
/test e2e-agent-sno-ipv4-pxe

zaneb commented:
/test ?

zaneb commented:
/test e2e-metal-assisted

cgwalters commented on data/data/bootstrap/files/etc/systemd/system/node-image-pull.service:
```diff
@@ -0,0 +1,21 @@
+After=sshd.service getty.target systemd-user-sessions.service
+
+[Service]
+# we use Type=notify so that systemd keeps retrying until we succeed. With
```
A useful trick...do you have a link for that?

zaneb replied:
Possibly this came from https://github.com/systemd/systemd/issues/1312#issuecomment-279172631
FWIW I attempted to use this in https://github.com/openshift/installer/pull/8911 and evidently it didn't work.
Consensus on the thread is that only https://github.com/systemd/systemd/issues/1312#issuecomment-228874771 works until we get systemd 254 and `RestartMode=direct`.

jlebon replied:
OK yes indeed, that doesn't work. I think this appeared to work when I was testing this but was actually for a different reason.

> Consensus on the thread is that only https://github.com/systemd/systemd/issues/1312#issuecomment-228874771 works until we get systemd 254 and `RestartMode=direct`.

Hmm, that requires having service B (the dependent service) also have `Restart=`. I think a cleaner workaround until `RestartMode=direct` is to just have in service A, `ExecStartPost=-systemctl start --no-block b.service`. Obviously in that case, you wouldn't have an `[Install]` section or a `Requires=`.

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+# XXX: this unit runs as unconfined_service_t and SELinux currently forbids the
+# transition to install_exec_t of ostree-container/rpm-ostree, which means we
+# can choke on labels unknown to the currently loaded policy
+setenforce 0
```
xref https://github.com/containers/bootc/blob/d91c00ab4ce33a20678abd9847125870f0f67b08/lib/src/lsm.rs#L132

jlebon replied:
`unconfined_service_t` is allowed to relabel itself, but when it reexecs itself, it still hits the same issue of `unconfined_service_t` not allowed the transition to `install_t`.

But you know who's allowed that transition? systemd. :)
So a simple workaround for this is actually:

```
ExecStartPre=chcon --reference=/usr/bin/ostree /usr/local/bin/node-image-pull.sh
```

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+
+# check out the commit
+checkout="${ostree_repo}/tmp/node-image"
+ostree checkout --repo "${ostree_repo}" ${hardlink} coreos/node-image "${checkout}" --skip-list=<(cat <<< "$etc_keep")
```
For example this I think is going to fail if the directory already exists and we retry the script.

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+# that's truly read-only
+ostree_repo=/ostree/repo
+hardlink='-H'
+if grep -q coreos.liveiso= /proc/cmdline; then
```
Ah yes, tricky.

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+    ostree_repo=/var/ostree-container/repo
+    mkdir -p "${ostree_repo}"
+    ostree init --mode=bare --repo="${ostree_repo}"
+    # use the system repo as parent to still get layer deduping
```
Wow, a usage of parent repos! I am curious if you tested and this works. So far its usage has been pretty obscure...

jlebon replied:
Yup, tested it manually and it worked fine! Haven't tested in a real live environment yet.

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+    ostree init --mode=bare --repo="${ostree_repo}"
+    # use the system repo as parent to still get layer deduping
+    ostree config --repo="${ostree_repo}" set core.parent /ostree/repo
+    # but we won't be able to force hardlinks cross-device
```
(of course, the thing is what we really want to do here is use composefs because actually that will let us seamlessly union together content in exactly scenarios like the live iso)

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+
+# Use ostree stack to pull the container here. This gives us efficient
+# downloading with layers we already have, and also handles SELinux.
+ostree container unencapsulate --repo "${ostree_repo}" --write-ref coreos/node-image \
```
Of everything, this is the most likely to fail and need retries; actually, it's really the *only* thing that should be retried in general right?

It may make sense to split this out.

Also...a very important and notable point: today `ostree container unencapsulate` will barf on non-ostree layers. I think we likely want `ostree container image pull` here (split out as a separate retry-able phase), then we can check out its ref.

zaneb commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+  ostree-unverified-image:docker://"${COREOS_IMAGE}"
+
+# immediately go back to enforcing
+setenforce 1
```
We have `set -e` enabled, so if the ostree command fails we will not re-enable selinux enforcing. I think you want to use something like `trap "setenforce 1" EXIT` as well.

cverna commented on data/data/bootstrap/files/etc/systemd/system/node-image-pivot.target:
```diff
@@ -0,0 +1,5 @@
+[Unit]
+Description=Node Image Pivot Target
```
Should we replace the term Pivot? We talked a bit about this but maybe this would be an opportunity to use the bootc terminology?

```suggestion
Description=Switch Node Image Target
```

jlebon replied:
Thinking of using the "overlay" terminology instead (e.g. `node-image-overlay.service`, "Node Image Overlay Target") since it's more accurate and distinct from switch/pivot which traditionally implied a reboot.

jlebon commented:
Just surfacing some insights from internal discussions here.

The AI test here is not using the ISO from `rhcos.json` but instead the one hardcoded in https://github.com/openshift/assisted-service/blob/master/data/default_os_images.json. 

The ABI tests here are not using the ISO from the `rhcos.json` but instead the ISO baked in the release payload within [`machine-os-images`](https://github.com/openshift/machine-os-images).

jlebon commented:
> The ABI tests here are not using the ISO from the `rhcos.json` but instead the ISO baked in the release payload within [`machine-os-images`](https://github.com/openshift/machine-os-images).

It's odd to me that `machine-os-images` is not more tightly coupled to `rhcos.json`. Was that by design? ISTM like `machine-os-images` should basically just be output from the installer repo too instead and always match what's in `rhcos.json`. (Though longer term obviously related is discussions around having managed bootimages, e.g. https://github.com/openshift/enhancements/pull/201.)

zaneb commented:
> It's odd to me that machine-os-images is not more tightly coupled to rhcos.json. Was that by design? ISTM like machine-os-images should basically just be output from the installer repo too instead and always match what's in rhcos.json.

It's tightly coupled in ART builds because machine-os-images depends on the installer image. But upstream CI doesn't have any notion of dependencies between container images. The CI payload just contains the latest master build of every container image. So machine-os-images only gets rebuilt when a PR merges to it, regardless of any changes to the installer repo (let alone unmerged PRs in the installer repo).

Arguably we chose the wrong ISO here, and in the event of a conflict between what rhcos.json says and what's in the payload, we should go with downloading the former so that we can CI changes like this. The assumption is that the payload is canonical, but that's not really true in CI, and CI is _theoretically_ the only place they can differ. (I wonder if this would break disconnected CI tests though - @bfournie do you know? It'd also mean we'd be executing different code paths in CI than users are actually getting in reality.)

r4f4 commented:
> Looks like e2e-aws-ovn-heterogeneous failed on a different issue.

This job is not required and expected to fail until https://github.com/openshift/installer/pull/8698 merges.

jlebon commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,66 @@
+
+# XXX: not sure why this is necessary given:
+# https://github.com/openshift/installer/blob/a3f5b449/pkg/asset/ignition/bootstrap/common.go#L411-L414
+chmod 0644 /etc/containers/registries.conf
```
Yeah, I saw that. Those patches are still there and specify 0644. Leaving debugging this bit in more details for later.

jlebon replied:
OK, narrowed this down to https://github.com/openshift/installer/pull/8977.

jlebon commented on data/data/bootstrap/files/usr/local/bin/node-image-pivot.sh:
```diff
@@ -0,0 +1,15 @@
+
+checkout="${ostree_repo}/tmp/node-image"
+
+mount -o bind,ro "${checkout}/usr" /usr
```
Yup, mentioned this already in the commit message:

> The big exception of course being the kernel. Relatedly, currently
`/usr/lib/modules` is also shadowed by the mount, but we could re-mount
it if needed.

jlebon commented:
OK, updated! Didn't rebase to make the diff easier, but also because I don't expect this to be merged before the associated enhancement is merged.

But changes now include:
- I renamed the nomenclature from "pivot" to "overlay"
- now with 100% less `setenforce 0`
- we no longer use `Restart=on-failure` because it complicates dependency handling (as discussed in https://github.com/openshift/installer/pull/8742#discussion_r1681857649), and only the tiny section related to networking needs to be retried
- I've split out the `systemctl isolate` back to `multi-user.target` into a separate unit; this is part of the work to make this work with Assisted Installer
- we now persist `/usr/lib/modules`
- we now use `ostree container image pull` to pull the image

This works with Assisted Installer as well but requires a patch there. I'll open that one shortly.

jlebon commented:
/test e2e-aws-ovn
/test e2e-openstack-ovn

jlebon commented:
OK, ended up doing a full rebase so try to get CI going.

/test e2e-aws-ovn
/test e2e-openstack-ovn

cgwalters commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,76 @@
+
+# massive hack to make ostree admin config-diff work in live ISO where /etc
+# is actually on a separate mount and not the deployment root proper... should
+# enhance libostree for this (remember, we're MountFlags=slave)
```
Right, in general a lot of shell script here that would probably be better moved into bootc at some point...something like `bootc switch --live` or so...

jlebon replied:
Yup, that's:

> Down the line, we should be able to replace some of this with
bootc apply-live once that's available (and also works in a live
environment). (See https://github.com/containers/bootc/issues/76.)

:)

cgwalters reviewed (approved):
Looks sane to me

jlebon commented:
Updated for feedback and updated commit and PR title to drop "pivot" wording.
Opened https://github.com/openshift/assisted-installer/pull/899 for Assisted Installer counterpart.

jlebon commented:
/test e2e-aws-ovn
/test e2e-openstack-ovn

jlebon commented:
One update I did here is in the live ISO case, we now nuke the temporary repo completely after the checkout. This saves precious RAM space so we can fit within the 16G minimum requirements for SNO.

/test e2e-aws-ovn
/test e2e-openstack-ovn

jlebon commented:
Rebased this now! Leaving it in draft for now, because I didn't retest things, but feel free to start reviewing.

/test e2e-aws-ovn
/test e2e-openstack-ovn

jlebon commented:
OK yup, still working fine! At least, testing in a local VM showed bootstrapping got far enough to tell me that the node overlay was successfully and `bootkube.sh` is continuing on its merry way. I'll let CI here make sure that the rest of the cluster install is successful, but otherwise taking this out of draft.

There's a DNM commit in this PR to ensure that CI tests run against bootimages . But once we're ready to merge this, I'll remove that commit.

One key thing to understand is that all the logic added in this PR is a no-op on the current bootimages which do ship OCP components. It only activates on the el-only bootimages. So we'll only be using this logic for real in that future bootimage bump where we decide to move over to the el-only bootimages.

jlebon commented:
/retest

jlebon commented:
/retest

patrickdillon commented:
/assign

mhrivnak commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,82 @@
+
+# Use ostree stack to pull the container here. This gives us efficient
+# downloading with layers we already have, and also handles SELinux.
+while ! ostree container image pull --authfile "/root/.docker/config.json" \
```
My assumptions / understanding just to confirm:

- the `$COREOS_IMAGE`  is an entire "bootc image" with a node layer, a coreos layer, and a rhel layer, as described in the enhancement proposal.
- at this point we're running on a basic RHEL image mode system, and we're using this command to retrieve `$COREOS_IMAGE` because we want to overlay its coreos layer and node layer on top of whatever RHEL we happen to be running right now.

Question: even though we only want the "node" and "coreos" layers, is it correct to assume that this command will pull the entire image including its rhel base image, unless that rhel base image happens to be the same as the one we're running at the moment?

jlebon replied:
> the `$COREOS_IMAGE` is an entire "bootc image" with a node layer, a coreos layer, and a rhel layer, as described in the enhancement proposal.

Yes, exactly!

> at this point we're running on a basic RHEL image mode system, and we're using this command to retrieve `$COREOS_IMAGE` because we want to overlay its coreos layer and node layer on top of whatever RHEL we happen to be running right now.

Nomenclature here gets messy, but basically I've been using:
- RHEL CoreOS image: RHEL layer + CoreOS layer
- node image: RHEL layer + CoreOS layer + node layer

The really confusing thing is that the node image in the payload is still called `rhel-coreos`, and I didn't have the stamina to also bundle as part of this enhancement renaming it to `node`, but ideally we eventually should rename so it's clearer.

Here, we're running on a RHEL CoreOS system (e.g. we still need Ignition for example). And the `COREOS_IMAGE` is the `rhel-coreos` component of the release payload (i.e. the node image).

> Question: even though we only want the "node" and "coreos" layers, is it correct to assume that this command will pull the entire image including its rhel base image, unless that rhel base image happens to be the same as the one we're running at the moment?

Think of the "layers" described in the enhancement as more conceptual than literal. In practice, the base RHCOS OCI images are chunked into many layers, grouping together files based on heuristics to try to maximize likelihood of sharing content. The node layer I guess is both conceptual and literal, since it all happens [in one `RUN` invocation](https://github.com/openshift/os/blob/155119346f2dd25801e9bf53b8741b0ba9790739/Containerfile#L35). So in the ideal case, we only pull that one node layer, which is currently around ~200M. The rest of the RHEL+CoreOS layers depend on how stale the bootimage is vs what we're installing (which I think more or less correlates with "how many z-stream releases have happened between the one containing this bootimage to the one we're installing").

Though I will note it'll be more inefficient than that at the start because the node image will still be built the old way while the bootimages have moved to the new model.

And of course, one can't not mention zstd:chunked here which should make it much easier to get efficiency than when playing with layers where we need to make sure everything is lined up just right (and notably that layer digests haven't been recalculated somewhere along the way in the pipelines).

jlebon replied:
> In practice, the base RHCOS OCI images are chunked into many layers, grouping together files based on heuristics to try to maximize likelihood of sharing content.

If you want an example of this BTW, you can look at the layers for any of the images in https://quay.io/repository/fedora/fedora-coreos?tab=tags.

mhrivnak commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,82 @@
+# check out the commit
+ostree checkout --repo "${ostree_repo}" ${hardlink} coreos/node-image "${ostree_checkout}" --skip-list=<(cat <<< "$etc_keep")
+
+# in the assisted-installer case, nuke the temporary repo to save RAM
```
How much RAM do we need to budget for this?

patrickdillon replied:
What is special about the assisted installer here? Does the assisted installer treat the bootstrap node separately? Perhaps this is mixed up with agent installer?

mhrivnak replied:
I assume it's because assisted-installer runs the "bootstrap in place" process from a live ISO, so the filesystem is in RAM.

jlebon replied:
> How much RAM do we need to budget for this?

Pretty much the size of the node image itself. I said ~1G when we discussed, but looking now, that's the compressed size. Uncompressed it's ~2.2G. I would add a healthy percent margin on top.

Worth mentioning I did test a SNO install with this in a 16G VM which is the minimum requirement. I'm not sure though how much headroom was left. I was actually pleasantly surprised when working on this to learn that bootstrapping needs less than 16G of container images. :)

Do you need more precise information?

> I assume it's because assisted-installer runs the "bootstrap in place" process from a live ISO, so the filesystem is in RAM.

Right, it's because we're booted from the ISO. We can't import into the system OSTree repo since it's read-only as part of the squashfs, so we have to create a separate OSTree repo and import there. That separate repo is in `/var`, which is in RAM.

Prashanth684 commented:
/test okd-scos-e2e-aws-ovn

Prashanth684 commented on data/data/coreos/rhcos.json:
```diff
@@ -1,107 +1,106 @@
 {
   "stream": "rhcos-4.16",
   "metadata": {
-    "last-modified": "2024-10-10T20:07:44Z",
```
need a similar scos.json file for centos bootimages.

jlebon replied:
Note this file is not going to be committed. It's just to exercise CI.

jlebon commented on data/data/bootstrap/files/usr/local/bin/node-image-pull.sh.template:
```diff
@@ -0,0 +1,82 @@
+# TimeoutStartSec=infinity.
+# This is fixed upstream by https://github.com/containers/common/pull/1758.
+# Should request backport...
+while ! COREOS_IMAGE=$(unset NOTIFY_SOCKET; image_for ${coreos_img}); do
```
Thanks, updated!

jlebon commented:
Updated to use `until` but also to retrigger CI!

Prashanth684 commented:
The OKD aws installs seems to be failing bootstrap. Unfortunately there are no journal logs and I couldn't connect to the instance when I tried locally.

Prashanth684 commented:
> The OKD aws installs seems to be failing bootstrap. Unfortunately there are no journal logs and I couldn't connect to the instance when I tried locally.

Ah...this is because it conflicts with the pivot that OKD does - so disabling that unit fixes this:

```
diff --git a/data/data/bootstrap/systemd/common/units/crio-configure.service.template b/data/data/bootstrap/systemd/common/units/crio-configure.service.template
index 1ae4af22cf..9113d6356a 100644
--- a/data/data/bootstrap/systemd/common/units/crio-configure.service.template
+++ b/data/data/bootstrap/systemd/common/units/crio-configure.service.template
@@ -2,9 +2,6 @@
 Description=Configure CRI-O to use the pause image
 After=release-image.service
 Requires=release-image.service
-{{if .IsOKD -}}
-Requires=release-image-pivot.service
-{{end -}}
 Before=crio.service
 
 [Service]
diff --git a/data/data/bootstrap/systemd/common/units/kubelet.service.template b/data/data/bootstrap/systemd/common/units/kubelet.service.template
index 8058ccbecf..092d4c8e6e 100644
--- a/data/data/bootstrap/systemd/common/units/kubelet.service.template
+++ b/data/data/bootstrap/systemd/common/units/kubelet.service.template
@@ -1,13 +1,7 @@
 [Unit]
 Description=Kubernetes Kubelet
 Wants=rpc-statd.service crio.service release-image.service
-{{if .IsOKD -}}
-Wants=release-image-pivot.service
-{{end -}}
 After=crio.service release-image.service
-{{if .IsOKD -}}
-After=release-image-pivot.service
-{{end -}}
 
 [Service]
 Type=notify
diff --git a/data/data/bootstrap/systemd/common/units/release-image-pivot.service.template b/data/data/bootstrap/systemd/common/units/release-image-pivot.service.template
index ebe833baf0..a7d7385668 100644
--- a/data/data/bootstrap/systemd/common/units/release-image-pivot.service.template
+++ b/data/data/bootstrap/systemd/common/units/release-image-pivot.service.template
@@ -4,6 +4,7 @@ Description=Pivot bootstrap to the OpenShift Release Image
 Wants=release-image.service
 After=release-image.service
 Before=bootkube.service kubelet.service dnsmasq.service
+Disabled=true
 
 [Service]
 Type=oneshot
```

jlebon commented:
@Prashanth684 Thanks! I've added a commit now which drops the OKD version of the pivot. Once this merges, I'll let you do a follow-up to clean things up and get rid of the FCOS bits in favour of SCOS (I guess bundled with the first SCOS bootimage bump; we can help you with that). It looked like there may be quite a few spots that will need to be adapted.

Prashanth684 commented:
/retest

jlebon commented:
CI failures appear unrelated and show up in other PRs.

patrickdillon commented:
/approve

sadasu commented:
/retest-required

jlebon commented:
Looks like this is nearing final approval and merge. I've removed the DNM commit that also changed `rhcos.json` since that was just for testing in CI.

andfasano commented:
/test ?

andfasano commented:
/test e2e-agent-compact-ipv4
/test e2e-agent-sno-ipv6
/test e2e-agent-ha-dualstack

travier commented on data/data/bootstrap/files/usr/local/bin/node-image-overlay.sh:
```diff
@@ -0,0 +1,15 @@
+rsync -a "${ostree_checkout}/usr/etc/" /etc
+
+# reload the new policy
+semodule -R
```
Should we `systemctl daemon-reexec` here to be safe?

jlebon replied:
We could, but I'd rather like to wait and see what kind of issues we hit before pre-emptively doing that. This may inform other things in this approach.

jlebon commented:
Just to export some conversations we had with @mhrivnak about this. To summarize a few points:
- This approach definitely has warts. After overlaying, we're going to still be running software from the bootimage, notably the network stack and systemd and obviously the kernel. The cleanest workaround for this would be to do a systemd soft-reboot (though in the ABI case, we do need to keep the agent alive across the transition). Obviously we can manually restart individual userspace components if the need arises.
- It's important to note that the bootimage and node images are both specified by the release payload (i.e. we still have and respect `rhcos.json`). So the skew is known and can be tested by CI, and it's the same skew that customers would run with.
- This change makes the release payload more "hermetic"; there's no longer two kubelet and cri-o versions to think about, only one. The contents of the bootimage still matter, but a lot less. This should in theory reduce bootimage churn as well (similarly to how e.g. traditional RHEL/Fedora install media is actually updated quite rarely after GA -- it just needs to work well enough to install and get updates).
- A middle-ground approach is to still produce bootimages with OCP content whenever needed. The CoreOS pipeline would still maintain mainly the RHEL X.Y-based streams, but we'd have a separate flow to build bootimages from the OCI node images using cosa/b-i-b whenever a bootimage bump is needed. This avoids having to change bootstrapping, but adds release engineering complexity and reduces turnaround speed.
- A lot of the gnarly bash code here should eventually become Rust code in bootc. The live application of updates is a desirable feature for RHEL Image Mode in general and something we'll want to support.
- @jlebon may be completely wrong and we may see a sea of red in CI dashboards (it wouldn't be after this merges, but after the bootimage bump where we switch to the new bootimages). If that happens, based on what fails and how, we may revert this change and go back to the drawing board.

jlebon commented:
> a clear way to know which coreos image and which node image were used.

I mentioned this in https://github.com/openshift/installer/pull/8742#issuecomment-2593605079, but the bootimage is specified by the payload as today. Of course, in some scenarios it's users that control the bootimage. The booted RHCOS version is visible in the journal logs but it's definitely not obvious. I've added a more direct journal log entry with that info now.

> a clear log message marking when the node image overlay happened (so when examining other log messages and events, we can know whether they happened with the old or new userspace)

Thanks, this is good feedback. Added more logging in general from the new scripts. Let me know if that matches what you were thinking!

https://github.com/openshift/release/pull/60534 is enabling unconditional collection of bootstrap logs, which should make all this easier to see in CI too but we'll of course only see it as part of the PR that changes the bootimages.

mhrivnak commented:
It's common for assisted-installer to boot a live ISO on systems before knowing which openshift version will get installed on them. That running live ISO will then be used for bootstrap-in-place once an installation is initiated.

@eranco74 @carbonin do we need anything else to accommodate that, or are we all set? Will we grab these scripts and unit files out of the bootstrap ignition?

@jlebon the main point otherwise is that we'll need flexibility to do the overlay on a variety of different versions of the coreos image.

jlebon commented:
> It's common for assisted-installer to boot a live ISO on systems before knowing which openshift version will get installed on them. That running live ISO will then be used for bootstrap-in-place once an installation is initiated.

Sorry yeah, you're right. I actually knew this (I even referenced the related late-binding problem in the commit message), but forgot it when writing my previous comment. That's actually a perfect example of "there's no longer two kubelet and cri-o versions to think about, only one".

I do have concerns though with not putting bounds on that skew to a range we can test and document. So I'd still argue for some version check as discussed in [this ticket](https://issues.redhat.com/browse/MGMT-16705).

> @eranco74 @carbonin do we need anything else to accommodate that, or are we all set? Will we grab these scripts and unit files out of the bootstrap ignition?

Yes. This is done by https://github.com/openshift/assisted-installer/pull/899.

mhrivnak commented on data/data/bootstrap/files/etc/systemd/system-generators/node-image-overlay-generator:
```diff
@@ -0,0 +1,9 @@
+
+UNIT_DIR="${1:-/tmp}"
+
+if ! rpm -q openshift-clients &>/dev/null; then
```
@eranco74 @carbonin this is the proxy test for whether the current filesystem needs all of the "node layer" stuff to be layered on. Any concerns for the assisted-installer live ISO? We don't install this, right?

carbonin replied:
No, we don't install anything additional on the rootfs

mhrivnak commented:
@jlebon ah yes, sorry I forgot you did that for us! Thanks for that.

The logging looks good to me. I think it'll help.

andfasano commented:
/test e2e-agent-compact-ipv4
/test e2e-agent-sno-ipv6
/test e2e-agent-ha-dualstack

andfasano commented:
I've tested locally via dev-scripts using also https://github.com/openshift/assisted-installer/pull/899 (and the related live boot ISO), and it went fine.

jlebon commented:
/retest

jlebon commented:
Updated this now! I've confirmed that https://github.com/containers/common/pull/1758 is in RHEL 9.6, so we can drop the workaround for it. Also switched from `bootc status` to `rpm-ostree status` for now to work around https://github.com/containers/bootc/issues/1043.

Tested both a standard flow and an ABI flow with this.

patrickdillon commented:
/lgtm
